src/engine/loopingcontrol.cpp

```diff
@@ -835,27 +923,104 @@ void LoopingControl::slotBeatLoop(double beats, bool keepStartPoint) {
         } else {
             newloopSamples.end += 2;
         }
-    } else if (newloopSamples.end > samples) {
+    }
+    if (newloopSamples.end > samples) {
         // Do not allow beat loops to go beyond the end of the track
         newloopSamples.end = samples;
+        // Let slotLoopHalve know to adjust beatloop_size
+        m_bBeatloopEndPointIsEndOfTrack = true;
+    } else {
+        m_bBeatloopEndPointIsEndOfTrack = false;
     }
 
-    if (keepStartPoint) {
+    if (keepStartPoint && (enable || m_bLoopingEnabled)) {
```

Comment:
Do we suffer a race condition, that the loop is left before it is enabled? 
Why is this only important for the "keepStartPoint" case?